### PR TITLE
fix(grpo): ReplayBuffer checkpointing and fix reservation leaks

### PR DESCRIPTION
**Pull request by macandro96**
# What does this PR do ?

- Adds replay-buffer checkpointing via `state_dict()`/`load_state_dict()`, saved as `replay_buffer.pt` alongside the dataloader state and restored on resume.

- Advance `last_target_weight_already_generated` in `sample()` (on batch consumption) instead of `add()`, and release per-target generation reservations only once all spawned workers complete. This prevents reservation leaks and premature target-skipping in the async collector.

## Note
Builds on top of https://github.com/NVIDIA-NeMo/RL/pull/2632 and hence it needs to be merged before this is merged.

# Issues
Fixes `increase last_target_weight_already_generated only when whole batch is generated and consumed` part of https://github.com/NVIDIA-NeMo/RL/issues/1906

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information on checkpointing

## Restore Behavior

On restore, the replay buffer is prepared for the current training step:
- drops trajectories for past target steps
- preserves current/future target steps so incomplete restored batches can be gap-filled
- resets `last_target_weight_already_generated` to `current_step - 1` so the collector re-checks current/future targets by count
- handles empty restored buffers without skipping the resumed step
- removes restored trajectories that are stale for their target step when `max_age_steps` is provided
- truncates after restore cleanup if current `max_size` is smaller than the checkpointed buffer

## Behavioral Differences From [Upstream Commit](https://github.com/NVIDIA-NeMo/RL/commit/0bb29f4452ca57d194138d0229915dd34dc4869e)

- Empty replay-buffer restore resets `last_target_weight_already_generated` to `current_step - 1`.
  - Reason: upstream only prepared the buffer when restored trajectories existed. If the restored buffer was empty but the checkpoint watermark was high, the collector could skip the resumed step and training would wait forever.

- Restore truncation happens after resume cleanup, not before.
  - Reason: upstream truncated immediately and kept the oldest entries. If `max_size` shrank, this could keep past-step trajectories and discard current/future ones needed for resume.

- When `current_training_step` is known, max-size truncation prioritizes the nearest current/future target steps.
  - Reason: after dropping past targets, the most useful restored trajectories are for the current training step first, then later steps. Keeping oldest-by-index is not necessarily the right behavior after resume.

- Restore can remove trajectories that are already stale for their target step via `max_age_steps`.
  - Reason: without this, restored stale trajectories could remain in the buffer until `sample()` evicts them later. Cleaning them during restore makes the restored state match what training can actually consume.

- GRPO passes `max_trajectory_age_steps` into replay-buffer restore.
  - Reason: restore needs the configured staleness window to remove stale restored trajectories consistently.

### CODE REVIEW

File: nemo_rl/algorithms/async_utils/trajectory_collector.py [hunk 1]
```diff
@@ -265,44 +305,117 @@ def _process_batch(self, batch: BatchedDataDict[DatumSpec]) -> None:
...
+            num_prompts_to_generate = min(num_prompts_in_batch, trajectories_needed)
+            if num_prompts_to_generate == 0:
+                print(
+                    f"🔄 Target {target_weight} already has enough trajectories, skipping"
+                )
+                with self._generation_check_lock:
+                    self._generating_targets.discard(target_weight)
+                return
+
+            if num_prompts_to_generate < num_prompts_in_batch:
```
Comment (yfw) [inline nemo_rl/algorithms/async_utils/trajectory_collector.py:322]: [`trajectory_collector.py:322`](https://github.com/NVIDIA-NeMo/RL/blob/de7bbd395da8040f1969e2c73620088b0e6fafa4/nemo_rl/algorithms/async_utils/trajectory_collector.py#L322)

**Coverage gap (reachable):** the partial gap-fill branch (`num_prompts_to_generate < num_prompts_in_batch`) is exercised on restore / after partial worker failure but has no collector-side test. The buffer-side `get_trajectories_needed` accounting is well covered; this branch that consumes it is not.

```python
    def test_process_batch_gap_fill_spawns_only_needed(self, monkeypatch):
        """Gap-fill generates only the trajectories still needed for a target."""

        class RemoteMethod:
            def __init__(self, value):
                self.value = value

            def remote(self, *args, **kwargs):
                return self.value

        class FakeReplayBuffer:
            def __init__(self):
                # Batch has 2 prompts, but only 1 more trajectory is needed.
                self.get_trajectories_needed = RemoteMethod(1)

        class FakeBatch:
            size = 2

            def slice(self, start, end):
                return self

            def repeat_interleave(self, repeats):
                return self

        started = []

        class RecordingThread:
            def __init__(self, *args, **kwargs):
                self._args = kwargs.get("args", ())

            def start(self):
                started.append(self._args)
                # Simulate the worker finishing and recording completion.
                target = self._args[2]
                with collector._counter_lock:
                    collector._completed_per_target[target] = (
                        collector._completed_per_target.get(target, 0) + 1
                    )

            def is_alive(self):
                return False

        target_weight = 7
        collector = self.create_local_collector(replay_buffer=FakeReplayBuffer())
        collector.running = True

        def reserve_target(generation_weight_version):
            collector._generating_targets.add(target_weight)
            return target_weight

        collector._get_next_target_for_generation = reserve_target
        monkeypatch.setattr(trajectory_collector_mod.ray, "get", lambda value: value)
        monkeypatch.setattr(
            trajectory_collector_mod._threading, "Thread", RecordingThread
        )

        collector._process_batch(FakeBatch())

        # Only one worker spawned even though the batch holds 2 prompts.
        assert len(started) == 1
        # Reservation released after spawning closed and the worker completed.
        assert target_weight not in collector._generating_targets
        assert target_weight not in collector._spawning_targets
        assert target_weight not in collector._spawned_per_target
        assert target_weight not in collector._completed_per_target
```

**Review by yfw — COMMENTED** [review]
## Team review — PR #2651

The replay-buffer **checkpointing** looks sound: `state_dict` / `load_state_dict` are symmetric, the snapshot is taken under the buffer lock, the state is saved alongside the other checkpoint artifacts inside `finalize_checkpoint`, and resume validates the checkpoint shape, drops past targets, evicts stale trajectories, keeps incomplete future targets for gap-fill, and truncates to `max_size` without dropping the current step's batch or creating target gaps. Restore from a pre-checkpointing run falls back to an empty buffer gracefully. It's well covered at the unit level (state_dict / load_state_dict round-trips, torch.save round-trip, gap-fill accounting, stale eviction, empty-state reset, max_size change, and validation errors).

Also fixes two real bugs in the async replay buffer: (1) premature reservation release (the old code released at `prompt_idx == 0` while later workers were still in flight, letting a subsequent batch re-reserve the same target and over-generate), and (2) premature watermark advancement in `add()` that could skip a partially-filled target and stall training.

@macandro96 Let's fold the close [#2632](https://github.com/NVIDIA-NeMo/RL/pull/2632) and just work on merging this one. Can you also add some docs to docs/guides/async-grpo.md about the replay buffer checkpointing since this is new behavior?

_Generated by Claude Code_

**Comment by macandro96** [comment]
> ## Team review — PR #2651
> The replay-buffer **checkpointing** looks sound: `state_dict` / `load_state_dict` are symmetric, the snapshot is taken under the buffer lock, the state is saved alongside the other checkpoint artifacts inside `finalize_checkpoint`, and resume validates the checkpoint shape, drops past targets, evicts stale trajectories, keeps incomplete future targets for gap-fill, and truncates to `max_size` without dropping the current step's batch or creating target gaps. Restore from a pre-checkpointing run falls back to an empty buffer gracefully. It's well covered at the unit level (state_dict / load_state_dict round-trips, torch.save round-trip, gap-fill accounting, stale eviction, empty-state reset, max_size change, and validation errors).
> 
> Also fixes two real bugs in the async replay buffer: (1) premature reservation release (the old code released at `prompt_idx == 0` while later workers were still in flight, letting a subsequent batch re-reserve the same target and over-generate), and (2) premature watermark advancement in `add()` that could skip a partially-filled target and stall training.
> 
> @macandro96 Let's fold the close [#2632](https://github.com/NVIDIA-NeMo/RL/pull/2632) and just work on merging this one. Can you also add some docs to docs/guides/async-grpo.md about the replay buffer checkpointing since this is new behavior?
> 
> _Generated by Claude Code_

Do you think its a good idea to add a function test for replay-buffer checkpointing? @yfw

**Comment by yfw** [comment]
> 
> Do you think its a good idea to add a function test for replay-buffer checkpointing? @yfw

Yes makes sense! If it is easy to do, would be good to have an e2e functional test

File: nemo_rl/algorithms/grpo.py [hunk 2]
```diff
@@ -2862,6 +2862,27 @@ def async_grpo_train(
         max_size=optimal_buffer_size
     )
 
+    last_checkpoint_path = checkpointer.get_latest_checkpoint_path()
+    if last_checkpoint_path is not None:
+        replay_buffer_path = os.path.join(last_checkpoint_path, "replay_buffer.pt")
```
Comment (terrykong) [inline nemo_rl/algorithms/grpo.py]: [`grpo.py:2867`](https://github.com/NVIDIA-NeMo/RL/blob/60fde000b4a36bdd17493c30593c29ab284e4c30/nemo_rl/algorithms/grpo.py#L2867)

**Test coverage gap (restore integration).** This restore branch — and the new `StubReplayBuffer.state_dict`/`load_state_dict` mocks added in `test_grpo.py` — are never exercised by the async unit tests, since they all force `checkpointer.get_latest_checkpoint_path()` to return `None`. The buffer-side `state_dict`/`load_state_dict` logic itself is well covered in `test_async_utils.py`; it's only the `async_grpo_train` wiring that's untested.

Here's a restore-wiring test (verified passing locally; it fails without this branch) you could add to `test_grpo.py`, inserting it right after `test_grpo_train_skips_reference_policy_logprobs_when_kl_disabled`:

```python
def test_async_grpo_train_restores_replay_buffer_from_checkpoint(
    mock_grpo_components, tmp_path
):
    """async_grpo_train restores the replay buffer when a checkpoint exists.

    Covers the restore wiring added to ``async_grpo_train`` (the
    ``get_latest_checkpoint_path`` -> ``replay_buffer.pt`` -> ``torch.load`` ->
    ``replay_buffer.load_state_dict`` path), which the other async tests skip
    because they force ``get_latest_checkpoint_path`` to return ``None``.
    """
    master_config = mock_grpo_components["master_config"]
    master_config.grpo["max_num_steps"] = 1
    master_config.grpo["max_num_epochs"] = 1
    master_config.grpo["val_period"] = 0
    master_config.grpo["val_at_start"] = False
    master_config.grpo["use_dynamic_sampling"] = False
    master_config.policy["generation"]["colocated"]["enabled"] = False

    # Write a real replay-buffer checkpoint next to the (fake) latest checkpoint.
    checkpoint_dir = tmp_path / "step_0"
    checkpoint_dir.mkdir()
    replay_buffer_path = checkpoint_dir / "replay_buffer.pt"
    restored_state = {
        "trajectories": [
            {"batch": {"data": f"restored_{i}"}, "rollout_metrics": {}}
            for i in range(3)
        ],
        "trajectory_versions": [0, 0, 0],
        "target_weight_versions": [0, 0, 0],
        "last_target_weight_already_generated": 0,
        "max_size": 10,
    }
    torch.save(restored_state, replay_buffer_path)

    checkpointer = mock_grpo_components["checkpointer"]
    checkpointer.get_latest_checkpoint_path.return_value = str(checkpoint_dir)

    grpo_save_state = _default_grpo_save_state()
    mock_rollout_metrics = {
        "mean_gen_tokens_per_sample": 10.0,
        "max_gen_tokens": 20,
        "min_gen_tokens": 5,
    }
    mock_batch = next(iter(mock_grpo_components["train_dataloader"]))
    policy = mock_grpo_components["policy"]

    real_torch_load = torch.load

    def _spy_load(path, *args, **kwargs):
        return real_torch_load(path, *args, **kwargs)

    with (
        mock_async_grpo_infrastructure(mock_batch, mock_rollout_metrics),
        _patched_logprob_phase(policy),
        patch(
            "nemo_rl.algorithms.grpo.torch.load", side_effect=_spy_load
        ) as mock_load,
    ):
        async_grpo_train(
            policy,
            None,
            mock_grpo_components["train_dataloader"],
            mock_grpo_components["val_dataloader"],
            mock_grpo_components["tokenizer"],
            mock_grpo_components["loss_fn"],
            mock_grpo_components["task_to_env"],
            mock_grpo_components["val_task_to_env"],
            mock_grpo_components["logger"],
            checkpointer,
            grpo_save_state,
            master_config,
        )

    # The restore branch must have read replay_buffer.pt from the latest checkpoint.
    loaded_paths = [call.args[0] for call in mock_load.call_args_list if call.args]
    assert str(replay_buffer_path) in loaded_paths, (
        "async_grpo_train did not restore the replay buffer from "
        f"{replay_buffer_path}; torch.load was called on {loaded_paths}"
    )
```

The save-side branch ([`state_dict` → `torch.save` of `replay_buffer.pt`](https://github.com/NVIDIA-NeMo/RL/blob/60fde000b4a36bdd17493c30593c29ab284e4c30/nemo_rl/algorithms/grpo.py#L3542)) remains uncovered — optional follow-up.

File: nemo_rl/algorithms/grpo.py [hunk 3]
```diff
@@ -2862,6 +2862,27 @@ def async_grpo_train(
...
+        if os.path.exists(replay_buffer_path):
+            print(f"📦 Restoring replay buffer from checkpoint: {replay_buffer_path}")
+            replay_buffer_state = torch.load(replay_buffer_path, weights_only=False)
```
Comment (terrykong) [inline nemo_rl/algorithms/grpo.py:2877]: [`grpo.py:2870`](https://github.com/NVIDIA-NeMo/RL/blob/60fde000b4a36bdd17493c30593c29ab284e4c30/nemo_rl/algorithms/grpo.py#L2870)

`weights_only=False` is correct and required here — trajectories are pickled `BatchedDataDict`/dicts, not plain tensors, so `weights_only=True` would reject them, and the checkpoint is a trusted same-job artifact. Consider a one-line comment so a future reader doesn't "harden" this to `weights_only=True` and silently break restore.

```suggestion
            # weights_only=False: trajectories are pickled BatchedDataDict/dicts,
            # not plain tensors. The checkpoint is a trusted same-job artifact.
            replay_buffer_state = torch.load(replay_buffer_path, weights_only=False)
```

File: tests/functional/L1_Functional_Tests_GRPO_2.sh [hunk 4]
```diff
@@ -38,6 +38,7 @@ run_test fast uv run --no-sync bash ./tests/functional/gdpo_async_grpo.sh
 run_test fast uv run --no-sync bash ./tests/functional/grpo_fsdp2.sh
 run_test      uv run --no-sync bash ./tests/functional/grpo_multiturn.sh
 run_test      uv run --no-sync bash ./tests/functional/grpo_non_colocated.sh
+run_test      uv run --no-sync bash ./tests/functional/grpo_async_replay_buffer_checkpoint.sh
```
Comment (terrykong) [inline tests/functional/L1_Functional_Tests_GRPO_2.sh]:
```suggestion
run_test fast uv run --no-sync bash ./tests/functional/grpo_async_replay_buffer_checkpoint.sh
```
we should set this test to fast so it runs on fast test set since this is a new test

**Review by terrykong — COMMENTED** [review]
## Team review — PR #2651

Reviewed by a team of agents (RL-code, bug-finder, test, comment-reviewer, devil's advocate). **The PR is in good shape — both bug fixes are confirmed genuine and no correctness bugs, deadlocks, races, or reservation leaks were found.**

- **Watermark move (`add()` → `sample()`)**: verified correct. Advancing `last_target_weight_already_generated` only on batch *consumption* (not on the first `add()`) fixes a real stall where a partially-filled target was permanently skipped. All live consumers now gate generation via `get_trajectories_needed` + the reservation set.
- **Deferred reservation release**: verified correct. The old `_process_batch` except path did no cleanup, so reserving a target then throwing before any worker spawned leaked the reservation forever. The new `_maybe_release_target` (release only when `completed >= spawned` and spawning has closed) closes this; a 500-trial concurrency stress test found 0 leaks, and lock ordering is safe (sequential, never nested).
- **Restore/cleanup** matches the documented behaviour, and the new [`async-grpo.md` Checkpointing section](https://github.com/NVIDIA-NeMo/RL/blob/60fde000b4a36bdd17493c30593c29ab284e4c30/docs/guides/async-grpo.md#L154) is accurate. The collector-side gap-fill branch flagged earlier now has a test (`test_process_batch_gap_fill_spawns_only_needed`). Linter (ruff/ruff-format/pyrefly) passes; the changed unit tests pass locally (58 + 69).
- **e2e functional test**: the PR now adds `tests/functional/grpo_async_replay_buffer_checkpoint.sh` (wired into `L1_Functional_Tests_GRPO_2.sh`), a two-phase save→restart→restore test. Its three log assertions match the real output (`grpo.py:2869`, `replay_buffer.py:337`, `grpo.py:3572`) — it looks sound and covers exactly the e2e gap that was open.

With the functional test in place, the remaining inline comments are minor **optional polish** (a one-line `weights_only=False` comment and a throughput note). Nothing blocks merge.

_Note: this builds on #2632, which should merge first._

_Generated by Claude Code_

**Review by yfw — APPROVED** [review]
